swift/ql/lib/codeql/swift/regex/internal/ParseRegex.qll

```diff
@@ -0,0 +1,1047 @@
+            // is an opening bracket (taking `^` into account)
+            exists(int posBeforePrevClosingBracket |
+              if this.getChar(prevClosingBracketPos - 1) = "^"
+              then posBeforePrevClosingBracket = prevClosingBracketPos - 2
+              else posBeforePrevClosingBracket = prevClosingBracketPos - 1
+            |
+              this.charSetDelimiter(index - 2, posBeforePrevClosingBracket) = true
+            )
+          then
+            // brackets without anything in between is not valid character ranges, so
+            // the first closing bracket in `[]]` and `[^]]` does not count,
+            //
+            // and we should _not_ mark the second opening bracket in `[][]` and `[^][]`
+            // as starting a new char set.                               ^           ^
+            exists(int posBeforePrevClosingBracket |
+              this.charSetDelimiter(index - 2, posBeforePrevClosingBracket) = true
+            |
+              result = this.charSetStart(posBeforePrevClosingBracket).booleanNot()
+            )
+          else
+            // if not, `pos` does in fact mark a real start of a character range
+            result = true
+        )
+      )
+    )
+  }
+
+  /**
+   * Helper predicate for chars that could be character-set delimiters.
+   * Holds if the (non-escaped) char at `pos` in the string, is the (one-based) `index` occurrence of a bracket (`[` or `]`) in the string.
+   * Result if `true` is the char is `[`, and `false` if the char is `]`.
+   */
+  boolean charSetDelimiter(int index, int pos) {
+    pos =
+      rank[index](int p |
+        (this.nonEscapedCharAt(p) = "[" or this.nonEscapedCharAt(p) = "]") and
+        // Brackets that art part of POSIX expressions should not count as
+        // char-set delimiters.
+        not exists(int x, int y |
+          this.posixStyleNamedCharacterProperty(x, y, _) and pos >= x and pos < y
+        )
+      ) and
+    (
+      this.nonEscapedCharAt(pos) = "[" and result = true
+      or
+      this.nonEscapedCharAt(pos) = "]" and result = false
+    )
+  }
+
+  /** Holds if a character set starts between `start` and `end`. */
+  predicate charSetStart(int start, int end) {
+    this.charSetStart(start) = true and
+    (
+      this.getChar(start + 1) = "^" and end = start + 2
+      or
+      not this.getChar(start + 1) = "^" and end = start + 1
+    )
+  }
+
+  /** Whether there is a character class, between start (inclusive) and end (exclusive) */
+  predicate charSet(int start, int end) {
+    exists(int innerStart, int innerEnd |
+      this.charSetStart(start, innerStart) and
+      not this.charSetStart(_, start)
+    |
+      end = innerEnd + 1 and
+      innerEnd =
+        min(int e |
+          e > innerStart and
+          this.nonEscapedCharAt(e) = "]" and
+          not exists(int x, int y |
+            this.posixStyleNamedCharacterProperty(x, y, _) and e >= x and e < y
+          )
```

Comment:
This looks like something that would be more efficiently expressed using `rank`, but if this doesn't perform poorly for other languages it's probably fine 👍.

Reply:
Good spot.  I'm reluctant to change this because (1) we don't currently have Swift test coverage for this edge case (you can remove the whole `not exists(...)` without any tests failing) and (2) I'm not sure what a clean solution with `rank` would look like.

Reply:
I've just added some new test cases around this stuff.  There are a couple of spurious parse failures around posix named character properties, but none of them are affected by this particular edge case (I'm struggling to think what would be).  The output of `parse.ql` LGTM.